extensions/interactions/RatioExpressionInput/directives/ratio-expression-input-validation.service.spec.ts

```diff
@@ -0,0 +1,231 @@
+    answerGroups[0].updateRuleTypesToInputs(
+      [isEquivalent, isEquivalentNonSimplified]);
+
+    warnings = validatorService.getAllWarnings(currentState,
+      customizationArgs, answerGroups, goodDefaultOutcome);
+    expect(warnings).toEqual([{
+      type: WARNING_TYPES.ERROR,
+      message: 'Rule 2 from answer group 1 will never be matched because' +
+      ' provided input is not in its simplest form.'
+    }, {
+      type: WARNING_TYPES.ERROR,
+      message: 'Rule 2 from answer group 1 will never be matched because' +
+      ' it is preceded by a \'IsEquivalent\' rule with a matching' +
+      ' input.'
+    }
+    ]);
+
+    let equalFourTerms = rof.createFromBackendDict({
+      rule_type: 'Equals',
+      inputs: {
+        x: [1, 2, 3, 4]
+      }
+    });
+
+    // The second rule will never get matched.
+    answerGroups[0].updateRuleTypesToInputs(
+      [hasNumberOfTermsEqualTo, equals, equalFourTerms]);
+
+    warnings = validatorService.getAllWarnings(currentState,
+      customizationArgs, answerGroups, goodDefaultOutcome);
+    expect(warnings).toEqual([{
+      type: WARNING_TYPES.ERROR,
+      message: 'Rule 3 from answer group 1 will never be matched because' +
+      ' it is preceded by a \'HasNumberOfTermsEqualTo\' rule with a matching' +
+      ' input.'
+    }]);
+
+    let equalsTwoTerms = rof.createFromBackendDict({
+      rule_type: 'Equals',
+      inputs: {
+        x: [1, 2]
+      }
+    });
+    let hasNumberOfTermsEqualToLength2 = rof.createFromBackendDict({
+      rule_type: 'HasNumberOfTermsEqualTo',
+      inputs: {
+        y: 2
+      }
+    });
+
+    // The second rule will never get matched.
+    answerGroups[0].updateRuleTypesToInputs(
+      [equalsTwoTerms, equals, hasNumberOfTermsEqualToLength2]);
+    warnings = validatorService.getAllWarnings(currentState,
+      customizationArgs, answerGroups, goodDefaultOutcome);
+    expect(warnings).toEqual([{
+      type: WARNING_TYPES.ERROR,
+      message: 'Rule 1 from answer group 1 will never be matched because' +
+      ' it has fewer number of terms than required.'
+    }, {
+      type: WARNING_TYPES.ERROR,
+      message: 'Rule 3 from answer group 1 will never be matched because' +
+      ' it has fewer number of terms than required.'
+    }, {
+      type: WARNING_TYPES.ERROR,
+      message: 'Rule 3 from answer group 1 will never be matched because' +
+      ' it is preceded by a \'HasNumberOfTermsEqualTo\' rule with a matching' +
+      ' input.'
+    }]);
+  });
+
+  it('should catch non-integer value for # terms', () => {
+    customizationArgs.numberOfTerms.value = 1.5;
+    var warnings = validatorService.getAllWarnings(
+      currentState, customizationArgs, answerGroups,
+      goodDefaultOutcome);
+    expect(warnings).toEqual([{
+      type: WARNING_TYPES.ERROR,
+      message: ('Number of terms must be integral.')
+    }]);
+  });
+
+  it('should catch undefined value for # terms', () => {
```

Comment:
we can't have an empty value because we have bounded type as a number for customization arg and for the number, there is no such thing as empty value.

Reply:
I could write a test for empty value but that would never be achieved until we make changes in backend and frontend both.

Reply:
updated